src/many/src/cose_helpers.rs

```diff
@@ -5,34 +5,84 @@ use coset::{
+        alg: Some(Algorithm::Assigned(coset::iana::Algorithm::EdDSA)),
+        key_ops,
+        params: [
+            vec![(
+                Label::Int(coset::iana::OkpKeyParameter::Crv as i64),
+                Value::from(coset::iana::EllipticCurve::Ed25519 as u64),
+            )],
+            params,
+        ]
+        .concat(),
+        ..Default::default()
+    }
+}
+
+/// Build an ECDSA CoseKey
+///
+/// # Arguments
+///
+/// * `(x, y)` - Public key
+/// * `d` - Private key
+pub fn ecdsa_cose_key((x, y): (Option<Value>, Option<Value>), d: Option<Value>) -> CoseKey {
+    // Allocate at least the public key
+    let mut params: Vec<(Label, Value)> = Vec::with_capacity(2);
+    let mut key_ops: BTreeSet<KeyOperation> = BTreeSet::new();
+
+    if let (Some(x), Some(y)) = (x, y) {
+        params.push((Label::Int(coset::iana::Ec2KeyParameter::X as i64), x));
+        params.push((Label::Int(coset::iana::Ec2KeyParameter::Y as i64), y));
+        key_ops.insert(KeyOperation::Assigned(coset::iana::KeyOperation::Verify));
+    }
+
+    if let Some(d) = d {
+        params.push((Label::Int(coset::iana::Ec2KeyParameter::D as i64), d));
+        key_ops.insert(KeyOperation::Assigned(coset::iana::KeyOperation::Sign));
+    }
+
+    // The CoseKeyBuilder is too limited to be used here
+    CoseKey {
+        kty: KeyType::Assigned(coset::iana::KeyType::EC2),
+        alg: Some(Algorithm::Assigned(coset::iana::Algorithm::ES256)),
+        key_ops,
+        params,
+        ..Default::default()
+    }
+}
+
 // TODO: Change the error type
 pub fn public_key(key: &CoseKey) -> Result<CoseKey, String> {
-    // The minicose version of this code discards any error and simply return None
     let params = BTreeMap::from_iter(key.params.clone().into_iter());
     match key.alg {
         Some(Algorithm::Assigned(coset::iana::Algorithm::EdDSA)) => {
             let x = params.get(&Label::Int(OkpKeyParameter::X.to_i64()));
-            if let Some(x) = x {
-                // The CoseKeyBuilder is too limited to be used here
-                let cose_key = CoseKey {
-                    kty: KeyType::Assigned(coset::iana::KeyType::OKP),
-                    alg: Some(Algorithm::Assigned(coset::iana::Algorithm::EdDSA)),
-                    key_ops: BTreeSet::from([KeyOperation::Assigned(
-                        coset::iana::KeyOperation::Verify,
-                    )]),
-                    params: vec![
-                        (
-                            Label::Int(coset::iana::OkpKeyParameter::Crv as i64),
-                            Value::from(coset::iana::EllipticCurve::Ed25519 as u64),
-                        ),
-                        (
-                            Label::Int(coset::iana::OkpKeyParameter::X as i64),
-                            x.clone(),
-                        ),
-                    ],
-                    ..Default::default()
-                };
-                Ok(cose_key)
+            if x.is_some() {
```

Comment:
After the change above, this can be simplified to `if let Some(x) = x.cloned() {`

Reply:
Fixed